src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java

```diff
@@ -0,0 +1,96 @@
+package org.jabref.logic.importer.fetcher;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.jabref.logic.importer.FulltextFetcher;
+import org.jabref.model.entry.BibEntry;
+
+import kong.unirest.Unirest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FulltextFetcher implementation that attempts to find a PDF URL at APS. Also see the <a
+ * href="https://harvest.aps.org/docs/harvest-api">API</a>, although it isn't currently used.
+ */
+public class ApsFetcher implements FulltextFetcher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ApsFetcher.class);
+
+    // The actual API needs either an API key or a header. This is a workaround.
+    private static final String DOI_URL = "https://www.doi.org/";
+    private static final String PDF_URL = "https://journals.aps.org/prl/pdf/";
+
+    @Override
+    public Optional<URL> findFullText(BibEntry entry) throws IOException {
+        Objects.requireNonNull(entry);
+
+        return entry.getDOI().map(doi -> {
+            String id;
+            try {
+                id = getId(doi.getDOI());
+            } catch (FileNotFoundException e) {
+                // No such DOI
+                return null;
+            } catch (IOException e) {
+                // Handle IOExceptions elsewhere
+                throw new UncheckedIOException(e);
+            }
+
+            String pdfRequest = PDF_URL + id;
+            int code = Unirest.get(pdfRequest).asJson().getStatus();
+
+            if (code == 200) {
+                LOGGER.info("Fulltext PDF found @ APS.");
+                try {
+                    return new URL(pdfRequest);
+                } catch (MalformedURLException e) {
+                    LOGGER.info("APS returned malformed URL, cannot find PDF.");
+                    return null;
+                }
+            } else {
+                return null;
+            }
+        });
```

Comment:
Please add `filter(result -> result != null)` to avoid returning an Optional containing a null value - or does this not happen? Did you add test cases for some of the `null` branches? ^^ - Maybe, you need to use [wiremock](http://wiremock.org/).